src/Spectre.libPeakFinder/ExtremaFinder.cpp

```diff
@@ -0,0 +1,100 @@
+/*
+ * ExtremaFinder.cpp
+ * Finding indices of extrema in signals.
+ *
+ Copyright 2018 Michal Gallus
+ 
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+ 
+ http://www.apache.org/licenses/LICENSE-2.0
+ 
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+#include "ExtremaFinder.h"
+#include "Spectre.libStatistics\Math.h"
+#include "Spectre.libFunctional\Filter.h"
+
+namespace spectre::algorithm::peakfinder
+{
+using namespace spectre::statistics::basic_math;
+using namespace spectre::core::functional;
+
+// Branchless cache-efficient zero filter
+static inline Indices GetNonZeros(const Signal& data)
+{
+    const size_t length = data.size();
+    std::vector<unsigned> indices(length + 1);
+    unsigned zeros = 0;
+
+    for (unsigned i = 0; i < length; i++)
+    {
+        bool dataIsZero = data[i] == 0.0;
```

Comment:
Comparison of `double` by means of `==`.

Reply:
That's fine, because in any float/double representation zero is represented by either all bits set to zero, or same, but the sign bit being set to one. Both cases are handled nicely though.

Reply:
Yeah, but what about potential problems with precision during numeric operations?

Reply:
If something would happen to not be equal to zero, then it'd either be slightly positive or negative (meaning the function is either growing or decaying), which is completely fine, and plays nicely with the rest of the algorithm

Reply:
The only operation that is applied on data in question is subtraction (using differentiate), so as long as the certain peaks have equal intensities there are no rounding errors.

Reply:
It'll for sure not make the sign of the result change, and this is the most critical part pretty much. Actually, perhaps it'd be an idea for future improvement to only store sign bits, instead of whole magnitudes.